### PR TITLE
chore(core-flows,inventory,types,medusa): add fixes to inventory module + location levels api

### PR DESCRIPTION
**Pull request by riqwan**
what:

- santizes inputs to prevent reserved_quantity from being updated directly
- inventory items create api can create location levels
- add validation to update quantity of reservation items
- general cleanup

RESOLVES CORE-2254

### CODE REVIEW

File: integration-tests/http/__tests__/inventory/admin/inventory.spec.ts [hunk 1]
```diff
@@ -612,6 +612,38 @@ medusaIntegrationTestRunner({
             })
           )
         })
+
+        it("should create inventory items along with location levels", async () => {
+          const response = await api.post(
+            `/admin/inventory-items?fields=%2blocation_levels.*`,
```
Comment (sradevski) [inline integration-tests/http/__tests__/inventory/admin/inventory.spec.ts]: I think you can just write 
```suggestion
            `/admin/inventory-items?fields=*location_levels`,
```

File: integration-tests/http/__tests__/reservations/admin/reservations.spec.ts [hunk 2]
```diff
@@ -88,16 +88,25 @@ medusaIntegrationTestRunner({
...
+          quantity: 10000,
           inventory_item_id: inventoryItem1.id,
           line_item_id: "line-item-id",
           location_id: stockLocation1.id,
         }
 
         const res = await api
-          .post(`/admin/reservations`, payload, adminHeaders)
+          .post(
+            `/admin/reservations?field=inventory_item.location_levels.*`,
```
Comment (sradevski) [inline integration-tests/http/__tests__/reservations/admin/reservations.spec.ts]: Similarly, the * goes in the beginning now

File: packages/modules/inventory-next/src/services/inventory-module.ts [hunk 3]
```diff
@@ -621,26 +670,47 @@ export default class InventoryModuleService<
...
 
+    const diff = arrayDifference(
+      ids,
+      reservationItems.map((i) => i.id)
+    )
+
+    if (diff.length) {
+      throw new MedusaError(
+        MedusaError.Types.INVALID_DATA,
+        `Reservation item with id ${ids} not found`
```
Comment (sradevski) [inline packages/modules/inventory-next/src/services/inventory-module.ts]:
```suggestion
        `Reservation item with id ${ids.join(', ')} not found`
```
Also, shouldn't we only print the diff?

File: packages/core/core-flows/src/inventory/workflows/create-inventory-items.ts [hunk 4]
```diff
@@ -3,23 +3,86 @@ import {
...
+  items: (InventoryNext.CreateInventoryItemInput & {
+    location_levels?: LocationLevelWithoutInventory[]
+  })[]
+}
+
+const buildLocationLevelMapAndItemData = (data: WorkflowInput) => {
+  const inventoryItems: InventoryNext.CreateInventoryItemInput[] = []
+  // Keep an index to location levels mapping to inject the created inventory item
+  // id into the location levels workflow input
+  const locationLevelMap: Record<number, LocationLevelWithoutInventory[]> = {}
```
Comment (sradevski) [inline packages/core/core-flows/src/inventory/workflows/create-inventory-items.ts:26]: I feel this index-based mapping will come to bite us in the butt at one point where the ordering of the operations isn't preserved 😄. I know we don't have much choice though, just thinking out loud.

File: packages/core/core-flows/src/inventory/workflows/create-inventory-items.ts [hunk 5]
```diff
@@ -3,23 +3,86 @@ import {
...
+  let index = 0
+
+  for (const { location_levels, ...inventoryItem } of data.items || []) {
+    locationLevelMap[index] = location_levels?.length ? location_levels : []
```
Comment (sradevski) [inline packages/core/core-flows/src/inventory/workflows/create-inventory-items.ts:29]:
```suggestion
    locationLevelMap[index] = location_levels ? location_levels : []
```

File: packages/core/core-flows/src/inventory/workflows/create-inventory-items.ts [hunk 6]
```diff
@@ -3,23 +3,86 @@ import {
...
+}
+
+const buildLocationLevelMapAndItemData = (data: WorkflowInput) => {
+  const inventoryItems: InventoryNext.CreateInventoryItemInput[] = []
+  // Keep an index to location levels mapping to inject the created inventory item
+  // id into the location levels workflow input
+  const locationLevelMap: Record<number, LocationLevelWithoutInventory[]> = {}
+  let index = 0
+
+  for (const { location_levels, ...inventoryItem } of data.items || []) {
```
Comment (sradevski) [inline packages/core/core-flows/src/inventory/workflows/create-inventory-items.ts]: Should we do `data.items.forEach` so you don't have to keep track of the index manually?

File: packages/core/core-flows/src/inventory/workflows/create-inventory-items.ts [hunk 7]
```diff
@@ -3,23 +3,86 @@ import {
...
+const buildInventoryLevelsInput = (data: {
+  locationLevelMap: Record<number, LocationLevelWithoutInventory[]>
+  items: InventoryNext.InventoryItemDTO[]
+}) => {
+  const inventoryLevels: InventoryNext.CreateInventoryLevelInput[] = []
+  let index = 0
+
+  // The order of the input is critical to accurately create location levels for
+  // the right inventory item
+  for (const item of data.items || []) {
```
Comment (sradevski) [inline packages/core/core-flows/src/inventory/workflows/create-inventory-items.ts]: Same here, maybe let's use forEach and get the index from there.

File: packages/core/core-flows/src/inventory/workflows/create-inventory-items.ts (hunk 4 above)
Comment (riqwan) [reply packages/core/core-flows/src/inventory/workflows/create-inventory-items.ts:26]: I'm on the same page as you are 😄 I really hate this, but there isn't an alternative here without the module accepting and returning a bulk identifier. Even in that case, you're still accepting the fact that the database is returning the right order its creating, which isn't guaranteed.

Comment (carlos-r-l-rodrigues) [reply packages/core/core-flows/src/inventory/workflows/create-inventory-items.ts:26]: Should we think about having a `reference_id` column where we control what is inserted and can 100% guarantee to map it back to the correspondent row?
maybe it is better to implement that now, before this happens in production

Comment (carlos-r-l-rodrigues) [reply packages/core/core-flows/src/inventory/workflows/create-inventory-items.ts:26]: or maybe `map_id`, something internal that we can always use in cases we need that mapping and there isn't any key yet

Comment (sradevski) [reply packages/core/core-flows/src/inventory/workflows/create-inventory-items.ts:26]: We will definitely need to consider something like that @carlos-r-l-rodrigues . Let me at least open a ticket, and we can revisit this after the public preview release.

Comment (riqwan) [reply packages/core/core-flows/src/inventory/workflows/create-inventory-items.ts:26]: Yes, thats the only way we can guarantee it

Comment (carlos-r-l-rodrigues) [reply packages/core/core-flows/src/inventory/workflows/create-inventory-items.ts:26]: but we generate the id of the entities in mikro-orm before inserting them. so there is a great chance that this is already being mapped correctly after inserting.

**Review by carlos-r-l-rodrigues — APPROVED** [review]

**Pull request merged**